### PR TITLE
 wait for jvm to listen before connecting debugger (#346)

### LINKED ISSUE #346: Error "Cannot connect" when restarting into debugging
follows up on #327 

steps:
1. ASSERT: have a server stopped
1. EXEC: start it into debug mode
1. ASSERT: server is running in debug mode
1. EXEC: restart it into debug mode

Result:
image
Vscode java debugger fails to connect to the server, server is stuck waiting for a debugger to connect. After some timeout the server gets back to "stopped"

### PR DESCRIPTION
**Pull request by adietish**
depends on #335 
fixes #346

### CODE REVIEW

**Comment by adietish** [comment]
this is basically done, needs the new client 0.16.0 to be published to npm first. Then it needs testing on OS'es other than Mac

**Comment by lstocchi** [comment]
> this is basically done, needs the new client 0.16.0 to be published to npm first. Then it needs testing on OS'es other than Mac

TestPR on other OSes

File: src/debug/debugInfo.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/*-----------------------------------------------------------------------------------------------
```
Comment (lstocchi) [inline src/debug/debugInfo.ts]: In the project you forgot to delete debugInfo.ts and debugInfoProvider outside the debug folder.

Comment (adietish) [reply src/debug/debugInfo.ts]: yes, rebasing error. thx for pointing out.

File: src/debug/javaDebugSession.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/*-----------------------------------------------------------------------------------------------
```
Comment (lstocchi) [inline src/debug/javaDebugSession.ts]: there are some tslist warnings (trailing whitespace mainly) here

File: src/debug/debugInfoProvider.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        const launchCommand: Protocol.CommandLineDetails = await client.getOutgoingHandler().getLaunchCommand({
+            mode: 'debug',
+            params: {
+                id: server.id,
+                serverType: server.type.id,
+                attributes: undefined
+            }
+        });
+
+        return this.create(launchCommand);
```
Comment (lstocchi) [inline src/debug/debugInfoProvider.ts]: maybe we should check launchCommand before to call the create method. Not sure if getLaunchCommand can return undefined or null

**Comment by lstocchi** [comment]
Tested on both Linux fedora and Windows. Works on both.

**Comment by adietish** [comment]
dont grock it, tests are all gree locally while the following 2 fail on travis:
```
  1) Command Handler
       restartServerInDebug
         should restart with given server:
     AssertionError: expected stub to have been called exactly once with arguments { mode: "debug", params: { attributes: [Map] {  }, id: "id", serverType: "type" } }
  	at Object.<anonymous> (out/test/commands.test.js:424:31)
  	at Generator.next (<anonymous>)
  	at fulfilled (out/test/commands.test.js:8:58)
  2) Command Handler
       restartServerInDebug
         should restart without given but prompted server:
     AssertionError: expected stub to have been called exactly once with arguments { mode: "debug", params: { attributes: [Map] {  }, id: "id", serverType: "type" } }
  	at Object.<anonymous> (out/test/commands.test.js:444:31)
  	at Generator.next (<anonymous>)
  	at fulfilled (out
```

ran ```npm install```, ```npm run build``` and also killed the ```out``` folder. It didnt help.

**Comment by adietish** [comment]
testPR

**Comment by adietish** [comment]
@lstocchi just reported that tests are green for him, too

**Comment by adietish** [comment]
turns out the tests failed if they were run in a fresh vm where no java extension were installed. @jrichter1 thx for spotting!

**Pull request merged**